server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java

```diff
@@ -0,0 +1,426 @@
+      {
+        if (!interval.contains(dataSegment.getInterval().getStartMillis())) {
+          throw new ISE("Failed to add segment to bucket: interval is not covered by this bucket");
+        }
+
+        // all values are pre-computed relatively to bucket start (which is considered as 0)
+        double t0 = convertStart(dataSegment, interval);
+        double t1 = convertEnd(dataSegment, interval);
+
+        double leftValue = FastMath.exp(t0) - FastMath.exp(t1);
+        double rightValue = FastMath.exp(-t1) - FastMath.exp(-t0);
+
+        SegmentAndValue segmentAndValue = new SegmentAndValue(dataSegment, leftValue, rightValue);
+
+        segments.tailSet(segmentAndValue).forEach(v -> v.leftValue += leftValue);
+        segments.headSet(segmentAndValue).forEach(v -> v.rightValue += rightValue);
+
+        SegmentAndValue lower = segments.lower(segmentAndValue);
+        if (lower != null) {
+          segmentAndValue.leftValue += lower.leftValue;
+        }
+
+        SegmentAndValue higher = segments.higher(segmentAndValue);
+        if (higher != null) {
+          segmentAndValue.rightValue += higher.rightValue;
+        }
+
+        segments.add(segmentAndValue);
+        return this;
+      }
+
+      public Builder removeSegment(DataSegment dataSegment)
+      {
+        SegmentAndValue segmentAndValue = new SegmentAndValue(dataSegment, 0.0, 0.0);
+
+        if (!segments.remove(segmentAndValue)) {
+          return this;
+        }
+
+        double t0 = convertStart(dataSegment, interval);
+        double t1 = convertEnd(dataSegment, interval);
+
+        double leftValue = FastMath.exp(t0) - FastMath.exp(t1);
+        double rightValue = FastMath.exp(-t1) - FastMath.exp(-t0);
+
+        segments.tailSet(segmentAndValue).forEach(v -> v.leftValue -= leftValue);
+        segments.headSet(segmentAndValue).forEach(v -> v.rightValue -= rightValue);
+        return this;
+      }
+
+      public boolean isEmpty()
+      {
+        return segments.isEmpty();
+      }
+
+      public Bucket build()
+      {
+        ArrayList<DataSegment> segmentsList = new ArrayList<>(segments.size());
+        double[] leftSum = new double[segments.size()];
+        double[] rightSum = new double[segments.size()];
+
+        int i = 0;
+        for (SegmentAndValue segmentAndValue : segments) {
+          segmentsList.add(i, segmentAndValue.dataSegment);
+          leftSum[i] = segmentAndValue.leftValue;
+          rightSum[i] = segmentAndValue.rightValue;
+          ++i;
+        }
+        return new Bucket(
+            new Interval(
+                interval.getStartMillis(),
+                segmentsList.get(segments.size() - 1).getInterval().getEndMillis()
+            ),
+            segmentsList,
+            leftSum,
+            rightSum
+        );
+      }
+    }
+  }
+
+  static class SegmentAndValue implements Comparable<SegmentAndValue>
+  {
+    private final DataSegment dataSegment;
```

Comment:
Maybe for sanity add equals() and hashCode() that compare the segment. Since we don't actually expect that the same segment is added, we may check
```java
if (!segments.add(segmentAndValue)) {
  throw new ISE("expect new segment");
}
```